statefun-sdk-js/src/storage.ts

```diff
@@ -0,0 +1,175 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+'use strict';
+
+import "./generated/request-reply_pb";
+import {TypedValueSupport} from "./types";
+import {Type, ValueSpec} from "./core";
+
+const M = global.proto.io.statefun.sdk.reqreply.FromFunction.PersistedValueMutation;
+
+const DEL = global.proto.io.statefun.sdk.reqreply.FromFunction.PersistedValueMutation.MutationType['DELETE'];
+
+// noinspection JSUnresolvedVariable
+const MOD = global.proto.io.statefun.sdk.reqreply.FromFunction.PersistedValueMutation.MutationType['MODIFY'];
+
+// noinspection JSValidateJSDoc
+class Value {
+    readonly #name;
+    readonly #type;
+    #box;
+    #mutated;
+    #deleted;
+
+    /**
+     *
+     * @param {string} name
+     * @param {Type} type
+     * @param {proto.io.statefun.sdk.reqreply.TypedValue} box
+     */
+    constructor(name: string, type: Type<any>, box: any) {
+        this.#name = name;
+        this.#type = type;
+        this.#box = box;
+        this.#mutated = false;
+        this.#deleted = false;
+    }
+
+    getValue<T>(): T | null {
+        if (this.#deleted) {
+            return null;
+        }
+        return TypedValueSupport.parseTypedValue(this.#box, this.#type);
+    }
+
+    setValue(jsObject: any) {
+        if (jsObject === undefined || jsObject === null) {
+            this.#mutated = true;
+            this.#deleted = true;
+            this.#box = null;
+        } else {
+            this.#mutated = true;
+            this.#deleted = false;
+            this.#box = TypedValueSupport.toTypedValue(jsObject, this.#type);
+        }
+    }
+
+    get name() {
+        return this.#name;
+    }
+
+    // internal helpers
+
+    asMutation(): any {
+        if (!this.#mutated) {
+            return null;
+        }
+        let mutation = new M();
+        mutation.setStateName(this.#name);
+        if (this.#deleted) {
+            mutation.setMutationType(DEL);
+        } else {
+            mutation.setMutationType(MOD);
+            mutation.setStateValue(this.#box);
+        }
+        return mutation;
+    }
+
+    static fromState(persistedValue: any, type: Type<any>) {
```

Comment:
Clearly persistedValue has a better type since you call a function on it. If persistedValue and type are related, can we do that through a generic?

Reply:
PersistedValue is a Protobuf generated class, but still I think that its a good type to parametrize `Value`, at least make it consistent internally.